cpp/arcticdb/async/async_store.hpp

```diff
@@ -404,65 +404,47 @@ class AsyncStore : public Store {
-            auto range = folly::Range(key_segments.begin() + start, key_segments.begin() + nxt);
-            auto key_segs = std::make_shared<std::vector<storage::KeySegmentPair>>();
-            auto key_seg_mutex = std::make_shared<std::mutex>();
-            std::vector<folly::Future<folly::Unit>> futs;
-            futs.reserve(range.size());
-
-            for (auto &[key, seg] : range) {
-                futs.emplace_back(async::submit_cpu_task(
-                    EncodeAtomTask(std::move(key),
-                                   ClockType::nanos_since_epoch(),
-                                   std::move(seg),
-                                   codec_,
-                                   count++,
-                                   encoding_version_))
-                                      .thenValue([de_dup_map, res](auto &&ks) {
-                                          auto key_seg = std::forward<decltype(ks)>(ks);
-                                          auto de_dup_key =
-                                              de_dup_map ? de_dup_map->get_key_if_present(key_seg.atom_key())
-                                                         : std::nullopt;
-
-                                          if (de_dup_key) {
-                                              ARCTICDB_DEBUG(log::version(),
-                                                             "Found existing key with same contents: using existing object {}",
-                                                             de_dup_key.value());
-                                              (*res)[key_seg.id()] = (de_dup_key.value());
-                                              return std::make_pair(storage::KeySegmentPair{}, false);
-                                          } else {
-                                              ARCTICDB_DEBUG(log::version(),
-                                                             "No existing key with same contents: writing new object {}",
-                                                             key_seg.atom_key());
-                                              (*res)[key_seg.id()] = (key_seg.atom_key());
-                                              return std::make_pair(std::move(key_seg), true);
-                                          }
-                                      }).thenValue([key_segs = key_segs, key_seg_mutex = key_seg_mutex](auto &&pk) {
-                    auto pair_key_write = std::forward<decltype(pk)>(pk);
-                    if (pair_key_write.second) {
-                        std::lock_guard kl{*key_seg_mutex};
-                        ARCTICDB_DEBUG(log::version(), "Enqueuing {} for write", pair_key_write.first.atom_key());
-                        key_segs->emplace_back(std::move(pair_key_write.first));
-                    }
-                }));
+        auto encode_futs = folly::window(key_segments, [*this] (auto&& ks) {
```

Comment:
This is more than just an addition of an `assert` 😆 

This is also in `batch_write` - how does this relate?